### PR TITLE
Digitizer mch

### PR DESCRIPTION
**Pull request by mwinn2**
Dear @aphecetche,

 I added now the timing from the hit.  I hope that this is what would be ok for the tracking.
If this is the case, I would actually prefer to do it always like that and never use the event-time (this is why I removed the member).  In this way, the processing of the hit array would be completely the same for continuous and triggered input std::vector <hits>.
The thing that needs to be clarified, if and if yes where the event-ID is actually needed. As far as I can see only for the MClabel, hence, this setter, getter is still required and would be properly filled in the Steer/DigitizerWorkflow/src/MCHDigitizerSpec.cxx
For the moment, I didn't remove yet setContinuous(bool val) and get... in Digitizer.h, but this could go, if it is correct  that we can get rid of the distinction for Digitizer.cxx

### CODE REVIEW

**Comment by mwinn2** [comment]
I did not find yet on the GUI of github the squash option, sorry for that, I will try to find it next time...

**Comment by mwinn2** [comment]
Alternatively, one could put the time of the event for every digit inside. 
But also this would not be different for continuous and non-continuous mode.

**Comment by aphecetche** [comment]
@mwinn2 I don't understand why there are so many commits : didn't you re-sync with dev before making the next changes ?

**Comment by aphecetche** [comment]
just a random thought. What if the "EventTime" in Digitizer be used as a "time offset" when merging events from different hit files ? One could imagine to take the hit times and offset them by different values depending on the input files. 
Must confess I've not checked how the merging is supposed to be done. But probably worth checking with other subsystems code and/or Sandro.

**Comment by mwinn2** [comment]
Ok, I will check with Sandro.
I thought I sync.ed, but apparently, the way I am doing it is not what is needed to be done.

**Comment by aphecetche** [comment]
> Ok, I will check with Sandro.
> I thought I sync.ed, but apparently, the way I am doing it is not what is needed to be done.

Well, usually the safest route is to acknowledge the short the life of a branch, the better. So that once your branch has been merged, you simply delete it and start a new one from the updated dev branch.

**Comment by mwinn2** [comment]
Hi Laurent,
ok, I checked with Sandro and in the TPC code. The addition makes sense. The units are in both cases ns. 
For the moment, the stepper is not filling a timing information. 
We could also just take the event time.

**Comment by mwinn2** [comment]
Hi @aphecetche ,
I added a member to make the "merging" of pile-up counts.
It is not yet completed and the usage of the multimap is not yet correct,(at least it compiles and works with dummy example however not yet tested in busy events) and it misses the MClabel handling, but before I continue, maybe you have a look, whether it is logically correct to do something like that or not or whether there is smarter way to spare time using another structure (I didn't see yet anything, since I don't see that the hits are ordered such that it becomes easier to do)

File: Detectors/MUON/MCH/Simulation/include/MCHSimulation/Digit.h [hunk 1]
```diff
@@ -31,7 +31,7 @@ class Digit : public DigitBase
  public:
   Digit() = default;
 
-  Digit(int pad, double adc);
+  Digit(double time, int pad, double adc);
```
Comment (aphecetche) [inline Detectors/MUON/MCH/Simulation/include/MCHSimulation/Digit.h:34]: I know this is not new, but I'm puzzled by the timestamp type. Why double  ? 
Probably more sth to see with Sandro, but I see there are two TimeStamp.h classes in O2 : 
- DataFormats/Headers/include/Headers/TimeStamp.h -> data oriented
- DataFormats/common/include/CommonDataFormat/TimeStamp.h -> used in all digitizers as far as I can tell.
The first one (data oriented) kind of makes sense to me as it's basically a long int. The second one, templatized, less so....

File: Detectors/MUON/MCH/Simulation/src/Digitizer.cxx [hunk 2]
```diff
@@ -157,12 +160,53 @@ int Digitizer::processHit(const Hit& hit, int detID, double event_time)
       q *= chargenon;
     }
     auto signal = resp.response(q);
-    digits.emplace_back(padid, signal);
+    digits.emplace_back(time, padid, signal);
+    multimap.emplace(padid, digits.size());
```
Comment (aphecetche) [inline Detectors/MUON/MCH/Simulation/src/Digitizer.cxx]: I would leave that forEachPadInArea loop untouched, and handle the digit merging logic only in the mergeDigits function, to keep things simple.

File: Detectors/MUON/MCH/Simulation/src/Digitizer.cxx [hunk 3]
```diff
@@ -157,12 +160,53 @@ int Digitizer::processHit(const Hit& hit, int detID, double event_time)
...
     ++ndigits;
   });
   return ndigits;
 }
+//______________________________________________________________________
+void Digitizer::mergeDigits(std::vector<Digit>& digits){
```
Comment (aphecetche) [inline Detectors/MUON/MCH/Simulation/src/Digitizer.cxx]: here I guess there are many ways to implement this. The choice should be performance- and readability- driven. 

That's one instance where we'd greatly benefit for a narrow unit test (well, unit benchmark in that case).
For instance generating fake digits with e.g. a known fraction of duplicates (fraction can then be varied from 0 to 1 to see how the implementation behaves), and then timing the merging method(s). 

Because as it is my guts tell me it's not efficient to use a multi map, a set and removal from a vector all at once. But my guts might well be totally wrong. Only measuring can tell.

**Comment by mwinn2** [comment]
Hi @aphecetche ,

thanks a lot for looking in detail! 
Ok, for separating the things w.r.t. the other loop.

I am probably busy for the rest of the afternoon, I will come up with a test set-up for benchmarking and then we see how to implement it best tomorrow at best.

For the time-stamp type, yes, to be understood why.

Michael

**Comment by mwinn2** [comment]
@aphecetche ,
- I debugged the piece of code that I wrote, it should at least do what is supposed to do.
I didn't manage with two items:
- for the timing test, I was wondering trying to write a simple test only for the merging, but somehow this didn't work with boost.
- for the inclusion of the test, I didn't get the CMake file modified such that I can probe both the geometry and the digitizer test that wrote.
For a different implementation, do you have a suggestion what to use instead of a multimap?

**Comment by mwinn2** [comment]
Hi @aphecetche ,
I made a new version of the merging in the branch digitizerMCHv2 without using a multimap, but only a map.
Let me know, if this is more reasonable in your view.

**Comment by aphecetche** [comment]
@mwinn2 will have a look this Friday probably.

**Comment by mwinn2** [comment]
Dear @aphecetche ,
I implemented the tests with the MClabels and the fasted version into the digitizer.
I have two open points from my side:
- I didn't get the boost testing for my digitizer working (yours works well) at the same time as the geometry, it is not very important, but I didn't find the proper way of using BOOST_AUTO_TEST_SUITE( etc. 
- the way I am using the std::vector objects with the labels and the digits is not really optimal. I tried a bit around, but maybe you can have a look at the mergeDigits function.
- I added two getters to get the two std::vector objects. 
In summary, maybe you check what you think about the current version that we can go ahead.
For the timing, I got in the end with what is inside now:
benchDigitMerging/0       1847 ns       1845 ns     389866
benchDigitMerging/1       1404 ns       1404 ns     495926
benchDigitMerging/2       5380 ns       5380 ns     123833

**Comment by mwinn2** [comment]
Hi @aphecetche ,
is there something that you see that i should do before opening the pull request?
Cheers
Michael

**Comment by aphecetche** [comment]
@mwinn2 A couple of things. 

What exactly is the problem with testing both digitizer and geometry ?

Also, I've just tried to run the tests : `ctest -R MCHSim` and it's a) very long b) failing... where a) could be  the reason for b... Do you really need a test that lasts for more than 100s or is the timeout the indication of a more severe issue ?  Anyway that has to be fixed before the PR.

And a minor thing : if we are to commit the different mergeDigits implementations tests (that might not be needed, except for the record maybe), I would have changed the names to something more meaningful than LA1, MW1 ;-)

**Comment by mwinn2** [comment]
H @aphecetche ,
concerning the first point: if I put into CMakeLists geometry as I would do (I am sure that there is something that escapes me), it doesn't compile together. I added the modification to this branch such that you can check what I do wrong. 
The output of the prompt during compilation looks like: 
-- FairRoot Version - nightly-20190330-37-g0ddb180 from - Fri Mar 29 16:05:07 2019 +0100
[3/4] Linking CXX executable bin/test_MCHSimulation
FAILED: bin/test_MCHSimulation 
: && /home/alidock/.sw/slc7_x86-64/GCC-Toolchain/v7.3.0-alice1-7/bin/c++  -fPIC -O2 -std=c++17 -O2 -g -DNDEBUG   Detectors/MUON/MCH/Simulation/test/CMakeFiles/test_MCHSimulation.dir/testDigitMerging.cxx.o Detectors/MUON/MCH/Simulation/test/CMakeFiles/test_MCHSimulation.dir/DigitMerging.cxx.o Detectors/MUON/MCH/Simulation/test/CMakeFiles/test_MCHSimulation.dir/testDigitization.cxx.o Detectors/MUON/MCH/Simulation/test/CMakeFiles/test_MCHSimulation.dir/testGeometry.cxx.o  -o bin/test_MCHSimulation -L/home/alidock/.sw/slc7_x86-64/GEANT3/v2-7-7/lib64  -L/home/alidock/.sw/slc7_x86-64/GEANT4/v10.4.2-3/lib  -L/home/alidock/.sw/slc7_x86-64/GEANT4_VMC/v3-6-p1-21/lib  -L/home/alidock/.sw/slc7_x86-64/vgm/v4-4-48/lib  -L/home/alidock/.sw/slc7_x86-64/ROOT/v6-16-00-7/lib  -L/home/alidock/.sw/slc7_x86-64/FairRoot/c672f280ec-1/lib  -L/home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib  -L/home/alidock/.sw/slc7_x86-64/pythia6/428-alice1-10/lib  -L/home/alidock/.sw/slc7_x86-64/pythia/v8235-alice2-1/lib -Wl,-rpath,/home/alidock/.sw/slc7_x86-64/GEANT3/v2-7-7/lib64:/home/alidock/.sw/slc7_x86-64/GEANT4/v10.4.2-3/lib:/home/alidock/.sw/slc7_x86-64/GEANT4_VMC/v3-6-p1-21/lib:/home/alidock/.sw/slc7_x86-64/vgm/v4-4-48/lib:/home/alidock/.sw/slc7_x86-64/ROOT/v6-16-00-7/lib:/home/alidock/.sw/slc7_x86-64/FairRoot/c672f280ec-1/lib:/home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib:/home/alidock/.sw/slc7_x86-64/pythia6/428-alice1-10/lib:/home/alidock/.sw/slc7_x86-64/pythia/v8235-alice2-1/lib:/home/alidock/.sw/BUILD/O2-latest/O2/lib:/home/alidock/.sw/slc7_x86-64/FairMQ/v1.3.8-5/lib:/home/alidock/.sw/slc7_x86-64/FairLogger/v1.2.0-7/lib /home/alidock/.sw/slc7_x86-64/googlebenchmark/1.3.0-7/lib/libbenchmark.a -lCore -lRIO -lGenVector -lCore -lRIO -lGenVector -lFairTools -lBase -lGeoBase -lParBase -lGeom -lCore -lVMC -lTree /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so -lBase -lFairTools /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_serialization.so -lpthread -lCore -lRIO -lGenVector -lFairTools -lBase -lGeoBase -lParBase -lGeom -lCore -lVMC -lTree /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so -lBase -lFairTools /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_serialization.so -lpthread -lEG -lPhysics -lCore -lRIO -lGenVector -lFairTools -lBase -lGeoBase -lParBase -lGeom -lCore -lVMC -lTree /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so -lBase -lFairTools /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_serialization.so -lpthread -lEG -lPhysics -lCore -lRIO -lGenVector -lFairTools -lBase -lGeoBase -lParBase -lGeom -lCore -lVMC -lTree /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so -lBase -lFairTools /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_serialization.so -lpthread -lEG -lPhysics /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so -lBase -lFairTools -lCore -lMathCore -lMatrix -lMinuit -lHist -lGeom -lGenVector -lRIO /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_container.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_container.so -lNet -lVMC -lGeom -lCore -lTree /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_iostreams.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_unit_test_framework.so lib/libMCHSimulation.so /usr/lib64/librt.so lib/libSimulationDataFormat.so lib/libDetectorsBase.so lib/libDataFormatsParameters.so lib/libField.so lib/libMathUtils.so -lNet lib/libCommonUtils.so lib/libReconstructionDataFormats.so lib/libCommonDataFormat.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_iostreams.so lib/libDataFormatsMID.so lib/libSimConfig.so lib/libDetectorsCommonDataFormats.so -lGeoBase -lParBase -lVMC -lTree /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_log_setup.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_serialization.so -lpthread lib/libMemoryResources.so /home/alidock/.sw/slc7_x86-64/FairMQ/v1.3.8-5/lib/libFairMQ.so.1.3.8 /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_program_options.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_thread.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_chrono.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_atomic.so /home/alidock/.sw/slc7_x86-64/FairLogger/v1.2.0-7/lib/libFairLogger.so.1.2.0 /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_date_time.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_filesystem.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_system.so /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_regex.so -lpthread -ldl -lrt /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_signals.so -lEG -lPhysics -lBase -lFairTools -lCore -lMathCore -lMatrix -lMinuit -lHist -lGenVector -lGeom -lRIO /home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/lib/libboost_container.so lib/libMCHMappingImpl3.so && :
Detectors/MUON/MCH/Simulation/test/CMakeFiles/test_MCHSimulation.dir/testGeometry.cxx.o: In function `main':
/home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/include/boost/test/unit_test.hpp:62: multiple definition of `main'
Detectors/MUON/MCH/Simulation/test/CMakeFiles/test_MCHSimulation.dir/testDigitization.cxx.o:/home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/include/boost/test/unit_test.hpp:62: first defined here
Detectors/MUON/MCH/Simulation/test/CMakeFiles/test_MCHSimulation.dir/testGeometry.cxx.o: In function `init_unit_test()':
/home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/include/boost/test/unit_test_suite.hpp:375: multiple definition of `init_unit_test()'
Detectors/MUON/MCH/Simulation/test/CMakeFiles/test_MCHSimulation.dir/testDigitization.cxx.o:/home/alidock/.sw/slc7_x86-64/boost/v1.68.0-8/include/boost/test/unit_test_suite.hpp:375: first defined here
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

**Comment by aphecetche** [comment]
You should ensure you have only one BOOST_TEST_MODULE and one BOOST_TEST_MAIN in the group of files that you use as the source of your test executable.

see https://www.boost.org/doc/libs/1_69_0/libs/test/doc/html/boost_test/utf_reference/link_references/link_boost_test_module_macro.html

**Comment by mwinn2** [comment]
@aphecetche ,
ok, thanks, this works.
For the fact that it stops, it seems to be related to the "Geometry" loading inside the thing (it had no problems, when I tested it last time), but I am still searching for the reason.

**Comment by mwinn2** [comment]
@aphecetche ,
I see the same slowliness problem, if I only test the Geometry (commenting the Digitizer-part in the CMakeLists.txt). To be sure, I did run the test with the dev in the repo, it is hanging, in the same way: 
Error in <TExMap::Add>: key 139637114664832 is not unique
Info in <TGeoManager::TGeoManager>: Geometry MCH-ONLY, ALICE MCH Standalone Geometry created
Info in <TGeoManager::SetTopVolume>: Top volume is cave. Master volume is cave
Info in <TGeoNavigator::BuildCache>: --- Maximum geometry depth set to 100
[INFO] No magnetic field found; using default tracking values 2 10 to initialize media


*** No errors detected
 hence, I think that it is not related to my code...

**Comment by mwinn2** [comment]
Do you have an idea what changed in geometry?

**Comment by aphecetche** [comment]
@mwinn2 well, testing on my side with the dev version I can run the testSimulation part just fine (in about 2s), meaning that the problem is located somewhere in this PR indeed

**Comment by mwinn2** [comment]
Hi @aphecetche ,
ok, I delete once more and try first with dev and then with my branch, what is the problem.

**Comment by aphecetche** [comment]
hum, actually retesting your branch I don't get the issue again ... Must have mixed up things on my side (probably used the wrong alienv)
But trying to see what was wrong, I noticed that the tests in testResponse is not included in the tests : 

```
~/alice/sw/BUILD/O2-latest-digitizerMCH/O2$ bin/test_MCHSimulation --list_content
o2_mch_simulation*
    DigitMergingIdentity*
        _0*
        _1*
        _2*
    DigitMerging*
        _0*
        _1*
        _2*
    geometrytransformer*
        CanGetAllChambers*
        GetRightNumberOfSlats*
        GetRightNumberOfQuadrants*
        GetDetElemVolumePath
        GetTransformations*
        TextualTreeDump*
    DigitizerTest*
    mergingDigitizer*
```

(which make sense as testResponse.cxx is not in the CMakeLists). Was there a reason for excluding this one ?

**Comment by mwinn2** [comment]
Hi @aphecetche ,
thanks a lot! my dev worked now as well with the testing, now I am recompiling my branch once more.
concerning Responsetesting: 
I didn't touch it for quite a while, so I didn't look into it, but there is no principle reason for not being included.

**Comment by mwinn2** [comment]
Hi @aphecetche ,
ok, it works now, I am currently modifying Responsetesting....

**Comment by mwinn2** [comment]
Hi @aphecetche ,
ok, I put now a working version of the Response inside. 
Please have a look as soon as you have time.
Michael

**Comment by mwinn2** [comment]
Hi @aphecetche ,
is there something obvious still missing from my side?  If not, I would go ahead to do a pull request.

Cheers
Michael
